### PR TITLE
add function to auto annotate images and generate object detection la…

### PR DESCRIPTION
**Pull request by Kayzwer**
…bels

### Summary
🔄🆕🧪

Added a new function to the `ultralytics.data.annotator` module for automatic detection and labeling of objects with YOLO. Updated the corresponding test function to use the new and renamed functions.

> _To annotate data with ease_
> _They refactored the module with these_
> _`auto_annotate_segment`_
> _And `auto_annotate_detect`_
> _For segmentation and YOLO, respectively_

### Walkthrough
*  Refactor `auto_annotate` function to separate segmentation and detection tasks ([link](https://github.com/ultralytics/ultralytics/pull/5133/files?diff=unified&w=0#diff-3e855cdd9b7d5f61f061a19aa1e4d05f63fb94b1dbbdfe75ddf9ef3db14796e0L8-R8), [link](https://github.com/ultralytics/ultralytics/pull/5133/files?diff=unified&w=0#diff-3e855cdd9b7d5f61f061a19aa1e4d05f63fb94b1dbbdfe75ddf9ef3db14796e0L22-R24), [link](https://github.com/ultralytics/ultralytics/pull/5133/files?diff=unified&w=0#diff-3e855cdd9b7d5f61f061a19aa1e4d05f63fb94b1dbbdfe75ddf9ef3db14796e0R51-R86))
* Update test function `test_data_annotator` to use new functions ([link](https://github.com/ultralytics/ultralytics/pull/5133/files?diff=unified&w=0#diff-51deb1ccb3a8b618784df2e2abfbc3b056436f1db72d1896d5102f9906978492L307-R310))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improve image annotation functionalities in Ultralytics software.

### 📊 Key Changes
- Removed support for a separate SAM segmentation model.
- Enhanced `auto_annotate` to support detection, segmentation, classify, and pose estimation using only the YOLO model.
- Created separate directories for each task output within the `auto_annotate_labels` folder.
- Included classification, detection, segmentation, and pose annotation in output files with corresponding labels and coordinates.
- Streamlined the codebase by consolidating multiple functionalities into the YOLO model.

### 🎯 Purpose & Impact
- **Single Model Simplicity**: The change simplifies the annotation pipeline by using one model, making it easier for users to perform annotations without managing multiple models.
- **Enhanced Organization**: Class-specific annotation directories improve the organization of annotated data, facilitating easier dataset management and use.
- **Multi-Task Support**: Supports multiple tasks like classification, object detection, segmentation, and pose estimation within the same framework, broadening the application scope of Ultralytics software.
- **Potential User Impact**: Users now have a more streamlined annotation experience with added capabilities, making it more convenient to generate datasets for varied machine learning tasks.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@Kayzwer this is a great idea! But I think we can simplify auto annotation here for all tasks in the existing function. You should be able to determine the result task by examining if the `segments` and `keypoints` fields are `not None` and then customizing the *.txt output appropriately.

**Comment by Kayzwer** [comment]
> @Kayzwer this is a great idea! But I think we can simplify auto annotation here for all tasks in the existing function. You should be able to determine the result task by examining if the `segments` and `keypoints` fields are `not None` and then customizing the *.txt output appropriately.
> 

That is a very good idea, and please give me some time to work on that

**Comment by glenn-jocher** [comment]
@Kayzwer hi there, thank you for your suggestion! It indeed sounds like a more streamlined and condensed approach to the annotation function. Distinguishing between different tasks based on whether the `segments` and `keypoints` fields are `None` or not, seems like a practical and efficient way to deal with multiple tasks. It allows the function to adapt the output depending on the type of task at hand. I appreciate your input and I will incorporate these ideas and work on it further. Please stay tuned for further updates.

**Comment by glenn-jocher** [comment]
@Kayzwer thanks for the updates! This looks better, but I wonder if we can use `result.save_txt()`?

**Comment by Kayzwer** [comment]
> @Kayzwer another option if we can't use `r.save_txt()` directly would be to re-use the `save_txt()` method code, which handles multi-task *.txt saving in a concise manner:
> 
> ```python
>     def save_txt(self, txt_file, save_conf=False):
>         """
>         Save predictions into txt file.
> 
>         Args:
>             txt_file (str): txt file path.
>             save_conf (bool): save confidence score or not.
>         """
>         boxes = self.boxes
>         masks = self.masks
>         probs = self.probs
>         kpts = self.keypoints
>         texts = []
>         if probs is not None:
>             # Classify
>             [texts.append(f'{probs.data[j]:.2f} {self.names[j]}') for j in probs.top5]
>         elif boxes:
>             # Detect/segment/pose
>             for j, d in enumerate(boxes):
>                 c, conf, id = int(d.cls), float(d.conf), None if d.id is None else int(d.id.item())
>                 line = (c, *d.xywhn.view(-1))
>                 if masks:
>                     seg = masks[j].xyn[0].copy().reshape(-1)  # reversed mask.xyn, (n,2) to (n*2)
>                     line = (c, *seg)
>                 if kpts is not None:
>                     kpt = torch.cat((kpts[j].xyn, kpts[j].conf[..., None]), 2) if kpts[j].has_visible else kpts[j].xyn
>                     line += (*kpt.reshape(-1).tolist(), )
>                 line += (conf, ) * save_conf + (() if id is None else (id, ))
>                 texts.append(('%g ' * len(line)).rstrip() % line)
> 
>         if texts:
>             Path(txt_file).parent.mkdir(parents=True, exist_ok=True)  # make directory
>             with open(txt_file, 'a') as f:
>                 f.writelines(text + '\n' for text in texts)
> ```

wow we actually have this function, I dont know XD

**Comment by Kayzwer** [comment]
but that function saves the prediction results instead of labels

**Comment by Kayzwer** [comment]
@glenn-jocher .save_txt() only produces object detection labels.

**Comment by glenn-jocher** [comment]
@Kayzwer no, .save_txt() saves all possible formats, including Segment and Keypoint results. That's the purpose of these lines in the function:

```python
                if masks:
                    seg = masks[j].xyn[0].copy().reshape(-1)  # reversed mask.xyn, (n,2) to (n*2)
                    line = (c, *seg)
                if kpts is not None:
                    kpt = torch.cat((kpts[j].xyn, kpts[j].conf[..., None]), 2) if kpts[j].has_visible else kpts[j].xyn
                    line += (*kpt.reshape(-1).tolist(), )
```

**Comment by Kayzwer** [comment]
@glenn-jocher the .save_txt() need to be modify 
image
the output string contain multiple task labels

**Comment by Kayzwer** [comment]
@glenn-jocher I completed the auto_annotate function to handle all kind of task, I also modify the test_auto_annotate in test_python.py to test different cases.

**Comment by glenn-jocher** [comment]
@Kayzwer That's fantastic to hear! Having the `auto_annotate` function handle different tasks will add a lot of versatility and functionality. Good job on modifying the `test_auto_annotate` in order to test different cases as well. Ensuring coded tests for various edge cases is vital for comprehensive testing and improves the robustness of the function. Please proceed with the next steps of your task. Thank you for your hard work and valuable contributions.

**Comment by Kayzwer** [comment]
@glenn-jocher Hi, do you see any ways to improve this piece of code?

**Comment by glenn-jocher** [comment]
@Kayzwer hi there, 

Of course, glad to help! I can definitely provide some general guidance to optimize your code. However, it's always beneficial to remember that improvements can be subjective and depend largely on the task at hand. 

1. **Code Readability**: Ensure your code is easy to read and understand. Well-commented code can be very beneficial for maintaining and debugging in the future. 

2. **Modular Approach**: Try to follow a modular approach where code is split into functions or modules. Each module or function should ideally have a single responsibility.

3. **Error Handling**: Don't forget to handle and manage exceptions properly. In the face of unexpected inputs your program should fail gracefully.

4. **Test Coverage**: Make sure you're covering as many cases as possible in your tests, including edge cases.

5. **Optimization and Efficiency**: Always look for opportunities to optimize your code and make it more efficient. 

Remember, programming is a continuous learning process and there is always room for refining and improving your code. Hope this helps!

**Comment by glenn-jocher** [comment]
@Kayzwer yes, the results objects have automatic txt saving built-in, so all the custom code in the PR is not required.

Also the SAM model has been removed, and can help with segmentation auto-annotation.

```python
from ultralytics import YOLO

# Load a model
model = YOLO('yolov8n.pt')

# Predict with the model
results = model()

# Save results to *.txt
results[12].save_txt('predict_results.txt')
```
